site/topics/testing/unittest.rst

```diff
@@ -152,6 +152,122 @@ Writing Unit Tests
+            self.assertEqual(2, sphere.diameter())
+
+        def test_diameter_radius_zero_arbitrary_centre_returns_zero(self):
+            sphere = Sphere(Point3D(1, 1, 1), 0)
+            self.assertEqual(0, sphere.diameter())
+
+        def test_diameter_radius_ten_arbitrary_centre_returns_twenty(self):
+            sphere = Sphere(Point3D(10, 11, 12), 10)
+            self.assertEqual(20, sphere.diameter())
+
+
+* Although there is nothing wrong with the above tests, we can instead, we can make use of ``subTest`` in this scenario
+
+
+.. code-block:: python
+    :linenos:
+    :emphasize-lines: 15, 16
+
+    import unittest
+
+    class SphereTest(unittest.TestCase):
+
+        # Other test methods not shown for brevity
+
+        def test_diameter_various_spheres_returns_correct_diameter(self):
+            cases = [
+                Sphere(Point3D(0, 0, 0), 0),
+                Sphere(Point3D(0, 0, 0), 1),
+                Sphere(Point3D(1, 1, 1), 0),
+                Sphere(Point3D(10, 11, 12), 10.1),
+            ]
+            expecteds = [0, 2, 0, 20.2]
+            for (case, expect) in zip(cases, expecteds):
+                with self.subTest(case=case, expect=expect):
+                    self.assertAlmostEqual(expect, case.diameter(), 5)
+
+
+* In the above example, each test input and expected output were stored in lists
+
+    * I used two separate lists, but there is nothing stopping you from using one list of tuples
+    * The variable names for the lists, ``cases`` and ``expecteds``, were arbitrary and by no means required
+
+* First notice the loop --- there is nothing particularly important for the ``subTest`` here, but the ``zip`` function has not been seen yet
+
+    * This just provides an easy way to loop over data within two lists at the same time
+    * This whole portion could be re-written as follows
+
+        .. code-block:: python
+            :linenos:
+
+            ...
+            for i in range(len(cases)):
+                with self.subTest(case=cases[i], expect=expecteds[i]):
+                    ...
+
+
+* Next, notice ``with self.subTest(case=case, expect=expect)``
+* There is a lot going on here, and much of it we will leave as *important code we need*
+
+* What is important is the assignment of the arguments
+
+    * ``case=case, expect=expect``
+
+* Up until now we have always specified the names and number of parameters a function/method takes
+* However, Python actually provides a way to provide an arbitrary number of arbitrarily named arguments
+* What is important is the name of the parameter we choose in these parentheses as they will be how we refer to the values
+
+    * ``self.assertAlmostEqual(expect, case.diameter(), 5)``
+    * This whole portion could be re-written as follows
+
+        .. code-block:: python
+            :linenos:
+
+            ...
+            with self.subTest(abc=case, xyz=expect):
+                self.assertAlmostEqual(abc, xyz.diameter(), 5)
```

Comment:
keyword args explained lightly here. I am hoping the example is enough to get the point across, but again, 👀